torch/csrc/distributed/rpc/ProcessGroupAgent.cpp

```diff
@@ -138,53 +132,56 @@ std::shared_ptr<FutureMessage> ProcessGroupAgent::send(
     future->markCompleted();
   }
 
-  enqueue(SendWork(dstRank, std::move(message)));
+  enqueueSend(RpcWork(dstRank, std::move(message)));
   return future;
 }
 
-void ProcessGroupAgent::enqueue(SendWork work) {
-  std::unique_lock<std::mutex> lock(sendQueueMutex_);
-  sendQueue_.emplace_back(std::move(work));
-  lock.unlock();
-
-  workProduceCV_.notify_one();
+void ProcessGroupAgent::enqueueSend(RpcWork work) {
+  threadPool_.run(std::bind(
+    [&](const RpcWork& work) {
+      std::stringstream ss;
+      serialize(work.message_, ss);
+      std::string str = ss.str();
+
+      std::vector<torch::Tensor> preamble = {
+        torch::tensor(
+          {
+            (int64_t)pg_->getRank(),
+            (int64_t)str.length(),
+          }, {torch::kLong})
+      };
+
+      // ProcessGroup is not thread-safe when sending with the same tag, hence
+      // the lock
+      std::unique_lock<std::mutex> lock(sendMutexes_.get()[work.rank_]);
+      pg_->send(preamble, work.rank_, work.rank_ /* channelTag */)->wait();
+      std::vector<torch::Tensor> payload =
+          {torch::from_blob((void *)str.c_str(), str.length(), {torch::kChar})};
+      pg_->send(payload, work.rank_, work.rank_ /* channelTag */)->wait();
+      lock.unlock();
```

Comment:
You can use a `std::lock_guard` as well, since you don't have to unlock it before it is destructed anyway.